spec/app/ics-028-cross-chain-validation/technical_specification.md

```diff
@@ -931,18 +958,78 @@ function onAcknowledgeVSCPacket(packet: Packet) {
   abortTransactionUnless(channelId IN channelToChain.Keys())
   chainId = channelToChain[packet.getDestinationChannel()]
 
-  stakingKeeper.UnbondValidators(chainID, packet.data.updates)
+  // iterate over the unbonding operations mapped to
+  // this chainId and vscId (i.e., packet.data.id)
+  foreach op in GetUnbondingOpsFromVSC(chainId, packet.data.id) {
+    // remove the consumer chain from 
+    // the list of consumer chain that are still unbonding
+    op.unbondingChainIds.Remove(chainId)
+    // if the unbonding operation has unbonded on all consumer chains
+    if op.unbondingChainIds.IsEmpty() {
+      // attempt to complete unbonding in Staking module
+      stakingKeeper.CompleteStoppedUnbonding(op.id)
+      // remove unbonding operation
+      unbondingOps.Remove(op.id)
+    }
+  }
+  // clean up vscToUnbondingOps mapping
+  vscToUnbondingOps.Remove((chainId, vscId))
 }
 ```
 - Initiator: 
   - The `onAcknowledgePacket()` method.
 - Expected precondition:
   - The IBC module on the provider chain received an acknowledgement of a `VSCPacket` on a channel owned by the provider CCV module.
 - Expected postcondition:
-  - The `UnbondValidators` method of the provider Staking module is invoked.
+  - For each unbonding operation `op` returned by `GetUnbondingOpsFromVSC(chainId, packet.data.id)`
+    - `chainId` is removed from `op.unbondingChainIds`;
+    - if `op.unbondingChainIds` is empty,
+      - the `CompleteStoppedUnbonding()` method of the Staking module is invoked;
+      - the entry `op` is removed from `unbondingOps`.
+  - `(chainId, vscId)` is removed from `vscToUnbondingOps`.
 - Error condition:
   - The ID of the channel on which the `VSCPacket` was sent is not mapped to a chain ID (in `channelToChain`).
 
+<!-- omit in toc -->
+#### **[CCV-PCF-GETUBDES.1]**
+```typescript
+// PCF: Provider Chain Function
+// Utility method
+function GetUnbondingOpsFromVSC(
+  chainId: Identifier, 
+  _vscId: uint64): [UnbondingOperation] {
+    ids = vscToUnbondingOps[(chainId, _vscId)]
+    if ids == nil {
+      // cannot find the list of unbonding operation IDs
+      // for this chainId and _vscId
+      return nil
+    }
+    // get all unbonding operations associated with
+    // this chainId and _vscId
+    ops = []
+    foreach id in ids {
+      // get the unbonding operation with this ID
+      op = unbondingOps[id]
+      // if cannot find UnbondingOperation according to vscToUnbondingOps,
+      // then vscToUnbondingOps was probably not correctly updated;
+      // programming error
+      abortSystemUnless(op != nil)
+      // append the operation to the list of operations to be returned
+      ops.Append(op)
+    }
+    return ops
+}
+```
+- **Initiator:** 
+  - The provider CCV module when receiving an acknowledgement for a `VSCPacket`.
+- **Expected precondition:**
+  - None. 
+- **Expected postcondition:**
```

Comment:
I wonder whether the pseudo code provides additional information that is not present in the expected postconditions? In general, is there is no algorithmic ideas, I have tried to reduce pseudo code as much as possible from specs. But perhaps that is against interchain standards?